KAFKA-4817; Add idempotent producer semantics

apurvam opened the pull request:
This is from the KIP-98 proposal. 

The main points of discussion surround the correctness logic, particularly the Log class where incoming entries are validated and duplicates are dropped, and also the producer error handling to ensure that the semantics are sound from the users point of view.

There is some subtlety in the idempotent producer semantics. This patch only guarantees idempotent production upto the point where an error has to be returned to the user. Once we hit a such a non-recoverable error, we can no longer guarantee message ordering nor idempotence without additional logic at the application level.

In particular, if an application wants guaranteed message order without duplicates, then it needs to do the following in the error callback:

1. Close the producer so that no queued batches are sent. This is important for guaranteeing ordering.
2. Read the tail of the log to inspect the last message committed. This is important for avoiding duplicates.

fpj committed:
Add broker producer id mapping support

guozhangwang committed:
KEOS: idempotent producer pid generation (#126)

Add the transaction coordinator for pid generation and management.

hachikuji committed:
Minor fixes for test failures and consistency (#133)

hachikuji committed:
Fix headers, new checkstyle rules, and some breakage in KafkaApis

hachikuji committed:
Avoid removal of non-expired PIDs when log cleaning (#138)

apurvam committed:
Client side implementation of the idempotent producer. (#129)

Commit:
Fix build errors due to rebase

hachikuji committed:
PIDs should be expired according to the transactional id expiration setting (#139)

Commit:
Fix build errors due to rebase

Commit:
Remove depependence of MemoryRecordsBuilder on TransactionState

hachikuji committed:
A few minor improvements (#150)

apurvam committed:
Reset transaction state on all irrecoverable exceptions (#153)

Commit:
Fix build errors due to rebase

apurvam commented:
cc @junrao

Commit:
Fix a todo, more renamings of LogEntry to RecordBatch

hachikuji commented on checkstyle/checkstyle.xml:
```diff
@@ -98,7 +98,7 @@
     <module name="ParameterNumber">
       <!-- default is 8 -->
-      <property name="max" value="12"/>
+      <property name="max" value="13"/>
```
I think we should consider turning off `ParameterNumber` check if we're just going to keep increasing it.

apurvam replied:
I think there is still some value in keeping this check. The reason the number is being bumped here is because the Sender constructor has added an argument. The solution would be to have a builder, but then a builder doesn't make sense for the Sender. 

It may make sense to just exempt the `Sender` for this check, but I am not sure if that is possible.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -0,0 +1,127 @@
+/**
+ * A class which maintains state for transactions. Also keeps the state necessary to ensure idempotent production.
+ */
+public class TransactionState {
```
The name seems like it could be a source of confusion. I wonder if we should rename this to something like `PidState` or `ProducerIdState` and maintain the actual transaction state separately? Do you think the coupling will be so tight that they will need to be tracked in the same class?

apurvam replied:
I don't know what a good name is for this. It currently maintains the `PidAndEpoch` and the PID->Sequence number mappings. Eventually, it will also store the transactional id, whether there is an active transaction, and the partitions belonging to the currently active transaction. 

There is no real coupling between the latter transactional state and the sequence number tracking except for the `PidAndEpoch`.  

As such, if we want to separate them, then a clean separation would require 3 classes. My preference would be to keep them all together, and just call it `TransactionState` or `TransactionalState`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -193,6 +193,22 @@ else if (maxTimestamp == RecordBatch.NO_TIMESTAMP)
+        if (isClosed() && (this.producerId != pid || this.producerEpoch != epoch || this.baseSequence != baseSequence)) {
+            // Sequence numbers are assigned when the batch is closed while the accumulator is being drained.
+            // If the resulting ProduceRequest to the partition leader failed for a retriable error, the batch will
+            // be re queued. When it is drained again, we expect it to receive the same sequence number as before.
```
I think this comment is out of date.

apurvam replied:
I think we should modify this check and throw an `IllegalStateException` if we try to set producer state after the batch is closed, as that should never happen with the current code.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/InitPidRequest.java:
```diff
@@ -0,0 +1,79 @@
+
+    public static class Builder extends AbstractRequest.Builder<InitPidRequest> {
+        private final String transactionalId;
+        public Builder(String transactionalId) {
```
I wonder if we should require the string to be null or non-empty. In Guozhang's current TC patch, we treat the empty string the same as null, but maybe we shouldn't actually allow the client to send an empty string? Seems doing so would be more likely to cause problems than not.

apurvam replied:
Yes. We should not allow empty strings IMO. My transactional producer patch treats an empty transactionalId as being 'unset', and I think it makes sense to enforce that across the board.

apurvam replied:
This is now fixed.

hachikuji reviewed (commented):
A few minor comments (I've seen all of this before obviously).

Commit:
Address PR Comments

apurvam commented:
The newly introduced integration test seems to be flaky on Jenkins. There are produce exceptions when bouncing brokers. I didn't see this locally, and I have run it 100's of times. Will look into it.

Commit:
Increase the number of retries in the ProducerBounceTest

hachikuji commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,366 @@
+   * Update the mapping to the given epoch and sequence number.
+   */
+  def update(pid: Long, pidEntry: PidEntry): Unit = {
+    if (pid >= RecordBatch.NO_PRODUCER_ID) {
```
This should either be `>` or `>= 0`. We could also move this check to the caller. Either way, we probably need a test case.

apurvam replied:
I have addressed this.

Commit:
In Sender.completeBatch, update the metadata everytime we get any type of InvalidMetadataException. With the recent refactor, we were only updating the metadata on non retriable exceptions

Commit:
Consolidate logic for testing whether a given producer id is valid. It
is now consistent everywhere.

apurvam commented:
I found the reason why the new integration test was failing. With a recent refactor of `Sender.completeBatch`, we only requested a metadata update on non-retriable errors. This was a regression, and as a result, when the leader for a partition changed, the metadata would not get updated and all the retries would get exhausted.

I moved the metadata update code to the correct place, and the tests are passing now. I also addressed other PR comments. cc @junrao

apurvam commented:
Somehow the new integration test only failed with JDK7 or scala 2.12, neither of which I run locally. So I never caught it until this PR.

Commit:
Remove extraneous new line.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -470,16 +611,33 @@ class Log(@volatile var dir: File,
 
+    val pidEntryMap = new mutable.HashMap[Long, PidEntryRange]()
     for (batch <- records.batches.asScala) {
-      // update the first offset if on the first message. For magic versions older than 2, we use the last offset
```
We seem to have lost this comment.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1259,6 +1264,33 @@ class KafkaApis(val requestChannel: RequestChannel,
+  }
+
+  def handleAddPartitionToTransactionRequest(request: RequestChannel.Request): Unit = {
+    throw new UnsupportedOperationException
```
Hmm, are we going to remove these methods before we merge this PR?

apurvam replied:
Yes.. I am about to push changes that removes these, here and in other places.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -68,6 +70,8 @@ class KafkaApis(val requestChannel: RequestChannel,
                 val clusterId: String,
                 time: Time) extends Logging {
 
+  val random = new Random()
```
This seems unused?

apurvam replied:
good catch. It was added during the initial implementations to generate PIDs before the server side code was ready.

Commit:
Fixing build errors after rebase onto trunk

hachikuji committed:
Refactor to pave way for LSO tracking (#154)

junrao commented on core/src/main/scala/kafka/coordinator/TransactionCoordinator.scala:
```diff
@@ -0,0 +1,92 @@
+  */
+object TransactionCoordinator {
+
+  def apply(config: KafkaConfig, zkUtils: ZkUtils, time: Time): TransactionCoordinator = {
```
It doesn't seem time is being used.

apurvam replied:
It seems to be called from `KafkaServer.scala:236`

junrao replied:
Apply() is being called, but is time actually being used?

apurvam replied:
Ah. So this is a reduced version of the class in the transactions branch. `time` is used in the `TransactionStateManager`, which is instantiated in the apply method.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -56,6 +56,11 @@ object LogAppendInfo {
+ *                   sequence of the first entry for the pid. The second PidEntry contains the last offset and last
+ *                   sequence for the last entry for the  pid. In the replication case, a single record could have multiple LogEntries
+ *                   from multiple pids. For a ProduceRequest to a leader, there will be one LogEntry (and hence one
+ *                   PID) per record.
```
Log entry => record batch.

Perhaps it's clearer to say records received in the follower, instead of replication.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -332,6 +353,62 @@ class Log(@volatile var dir: File,
+    *
+    * @param newStartOffset New start offset to clean up the map
+    */
+  private[log] def updateIdMap(newStartOffset: Long): Unit = {
```
Hmm, not sure why we need to expire the ids before the dirty offset.

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+  }
+
+  private def formatDirName(topicPartition: TopicPartition): String = {
+    s"$DirnamePrefix-${topicPartition.topic}-${topicPartition.partition}"
```
It seems that the snapshots are created under a dir named topic-partition. So, it seems we don't need to include ${topicPartition.topic}-${topicPartition.partition} here?

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+      case None =>
+        lastSnapOffset = -1
+        lastMapOffset = -1
+        snapDir.mkdir()
```
Is this needed since we already did that during initialization of the class?

apurvam replied:
Good catch. Deleted.

apurvam replied:
Actually, I undeleted these lines as they are needed. The `loadSnapshot` can be called in two cases: During initial start, and also when the log is truncated. In the latter case, if there is no previous snapshot, we need to reset to the start offset, which is what this code does.

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+    }
+  }
+
+  def truncateAndReload(logEndOffset: Long) = {
```
It doesn't seem that we are returning a value. If so, we want to remove =.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1308,6 +1313,17 @@ class KafkaApis(val requestChannel: RequestChannel,
+    // Send response callback
+    def sendResponseCallback(result: InitPidResult): Unit = {
+      val responseBody: InitPidResponse = new InitPidResponse(result.error, result.pid, result.epoch)
+      trace(s"Generated new PID ${result.pid} from InitPidRequest from client ${request.header.clientId}")
```
Probably log initPidRequest too?

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1308,6 +1313,17 @@ class KafkaApis(val requestChannel: RequestChannel,
+      trace(s"Generated new PID ${result.pid} from InitPidRequest from client ${request.header.clientId}")
+      requestChannel.sendResponse(new RequestChannel.Response(request, responseBody))
+    }
+    txnCoordinator.handleInitPid(initPidRequest.transactionalId, sendResponseCallback)
```
We will need to check if the request is authorized.

apurvam replied:
I added a JIRA to track this in a future PR: https://issues.apache.org/jira/browse/KAFKA-4970

apurvam commented:
Current TODOS from discussing with Jun offline:

1. Filed https://issues.apache.org/jira/browse/KAFKA-4970 as a future improvement. 
2. Need to remove the `Log.updateIdMap` which removes entries before the current dirty offset from the pid map. this is a bug which would result in real entries being lost. 
3. Need to name the snapshot files appropriately. Currently they are in the partition directory, but have the topic partition in their name, which is redundant. 
4. Need to ensure that `ProducerIdMapping.maybeTakeSnapshot` is called periodically.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -364,10 +441,10 @@ class Log(@volatile var dir: File,
    */
   def append(records: MemoryRecords, assignOffsets: Boolean = true): LogAppendInfo = {
-    val appendInfo = analyzeAndValidateRecords(records)
+    val appendInfo = analyzeAndValidateRecords(records, isFromClient = assignOffsets)
```
In ReplicaFetcherThread, we probably want to log a warning if LogAppendInfo.isDuplicate is true after the append() call since it's not expected.

apurvam replied:
I have added this log line.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -497,17 +582,21 @@ class Log(@volatile var dir: File,
-      // to avoid the need to decompress the data (the last offset can be obtained directly from the wrapper message).
-      // For magic version 2, we can get the first offset directly from the batch header.
+      if (isFromClient && batch.magic >= RecordBatch.MAGIC_VALUE_V2 && shallowMessageCount > 0)
+        throw new InvalidRecordException("Client produce requests should not have more than one batch")
```
Do we need this? It seems that we already validate this in ProduceRequest.validateRecords() when the broker receives the produce request.

apurvam replied:
Not sure I follow. The `ProduceRequest.validateRecords` just checks that the right message format goes with the right version. This check further validates that there is exactly one `RecordBatch` in a produce request with the new message format.

Seems to me that the checks complement each other.

junrao replied:
ProduceRequest.validateRecords() has the following, right?

```
            MutableRecordBatch entry = iterator.next();
            if (entry.magic() != RecordBatch.MAGIC_VALUE_V2)
                throw new InvalidRecordException("Produce requests with version " + version + " are only allowed to " +
                        "contain record batches with magic version 2");

            if (iterator.hasNext())
                throw new InvalidRecordException("Produce requests with version " + version + " are only allowed to " +
                        "contain exactly one record batch");

```

apurvam replied:
You are right. Will delete this check.

apurvam replied:
So I thought about this a bit more. I think it still make sense to validate this on the server side. I assume that the librdkafka clients may not have the request side validation, so it would be good to have server side validation before we write to the log.

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+
+  private def validateAppend(epoch: Short, firstSeq: Int, lastSeq: Int) = {
+    if (this.epoch > epoch) {
+      throw new ProducerFencedException(s"Invalid epoch (zombie writer): $epoch (request epoch), ${this.epoch}")
```
Could we make it clear that the latter epoch is the server epoch?

apurvam replied:
It is actually the producer epoch which is stale in this case, but will clarify the exception message.

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+
+  def checkForExpiredPids(currentTimeMs: Long) {
+    pidMap.retain { case (pid, lastEntry) =>
+      currentTimeMs - lastEntry.timestamp < maxPidExpirationMs
```
Should we do the same check for expiration when loading a snapshot?

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+    *
+    * The parent directory is a concatenation of:
+    *   - log.dir
+    *   - base name
```
What is a base name?

apurvam replied:
I think this comment is outdated. I updated with the actual logic. The snapshot files will be located inside a `pid-mapping` subdirectory of the log directory. The files themselves will be named with the pattern `$lastOffset.snapshot`.

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+    *
+    *   The offset is supposed to increase over time and
+    */
+  def maybeTakeSnapshot() {
```
It doesn't seem that we have the logic to take snapshots periodically since this method is only called from tests?

apurvam replied:
Good catch, will add a periodic cleaner task.

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -568,6 +574,11 @@ object KafkaConfig {
+  val TransactionIdExpirationMsDoc = "The maximum time of inactivity before a transactional id is expired by the " +
+    "transaction coordinator. Note that this also influences PID expiration: PIDs are guaranteed to expire " +
+    "after expiration of this timeout from the last write by the PID (they may expire sooner if the last write " +
+    "from the PID is deleted after due to the topic's retention settings)."
```
remove after?

junrao reviewed (commented):
@apurvam : Thanks for the patch. Some comments on the server side.

Commit:
Don't clean the pidMap from the log cleaner. We used to clean all
entries before the dirty entry, but this would have gotten rid of some
valid pid entries, and is a bug.

apurvam committed:
First system test for the idempotent producer (#155)

Modified the replication test to check for duplicates if idempotence is enabled.

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -334,26 +301,84 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+                        "enabled, the producer tracks additional metadata to ensure that messages are deduplicated " +
+                        "on the broker. Without retries, there will never be duplicates to begin with.");
+            }
+            if (config.getInt(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION) != 1) {
```
It would be inconvenient for a user to have to configure 2 other properties after enabling idempotence. Perhaps we could set these 2 values to a reasonable default (e.g., 3 retries) if the user doesn't configure these properties explicitly. If the user explicitly set those properties with an incorrect value, we can then throw an exception.

apurvam replied:
I have implemented this. The silght modification is that since the only invalid retries config is 0, which is the default, I just override the default to 3 when idempotence is enabled.

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +282,50 @@ public void forceClose() {
+            try {
+                Node node = getReadyNode(requestTimeout);
+                if (node != null) {
+                    ClientResponse response = sendAndAwaitInitPidRequest(node);
```
Hmm, in sendAndAwaitInitPidRequest(), we just initiate a pid request w/o checking if we actually can send to the node. Not sure if this is safe. NetworkClient also sends internal metadata request. So, it's possible that an ongoing metadata request is still pending on the same node and the send of pid request will hit an IllegalStateException in NetworkClient.doSend().

apurvam replied:
Synced offline, and we agreed this isn't a real problem since both `getReadyNode` and `sendAndAwaitInitPidRequest` call poll. This will mean than any outstanding requests will get a chance to be processed there will be no deadlock.

junrao reviewed (commented):
A couple of initial comments on the producer side.

hachikuji commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -167,9 +168,17 @@ def validate(self):
             msg = self.annotate_data_lost(data_lost, msg, len(to_validate))
 
 
+        if self.enable_idempotence is True:
```
nit: `is True` seems redundant.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -460,6 +485,7 @@ private static int parseAcks(String acksString) {
             // first make sure the metadata for the topic is available
             ClusterAndWaitTime clusterAndWaitTime = waitOnMetadata(record.topic(), record.partition(), maxBlockTimeMs);
             long remainingWaitMs = Math.max(0, maxBlockTimeMs - clusterAndWaitTime.waitedOnMetadataMs);
+            remainingWaitMs = maybeWaitOnPid(remainingWaitMs);
```
Discussed offline. We agreed it's not actually necessary to block here since the sequence number is not assigned until the batch is ready to be sent anyway.

apurvam replied:
I just deleted this block from the producer.

hachikuji commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+    ProducerIdEntry(epoch, lastSeq, lastOffset, lastSeq - firstSeq + 1, lastTimestamp)
+}
+
+private[log] case class PidEntryRange(first: ProducerIdEntry, last: ProducerIdEntry)
```
This is no longer used.

hachikuji commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,387 @@
+      throw new IllegalArgumentException(s"Unhandled snapshot file version $version")
+
+    val crc = struct.getUnsignedInt(CrcField)
+    if (crc != Crc32.crc32(buffer, PidEntriesOffset, buffer.length - PidEntriesOffset))
```
Now that we've merged the CRC32C patch, we may as use that.

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -225,11 +230,20 @@ public double compressionRate() {
 
     public boolean isFull() {
-        return recordsBuilder.isFull();
+        return !isWritable || recordsBuilder.isFull();
```
Not sure if we strictly needs to isWriteable. Currently, during append(), if the current producerBatch is full, we just create a new batch.

apurvam replied:
Hmm. You may be right. The reason I introduced `isWritable` is because we can no longer close a batch when it is full. we can only close it at the point of sending in order to set the right sequence number. So I introduced `isWritable` to denote the state where it can no longer take appends, but is not closed. 

Your point is valid: once a batch is considered full for a particular append, no future appends should go to it since there will be another batch at the tail which should get the new appends.

apurvam replied:
So, looking over it a bit more, I now know why I introduced `isWritable`. The `isFull` method is not only used during append. It is also used to wake the sender up: if the deque is of size one and the only batch in there is not full, the sender will not be woken up to drain until it hits the linger ms. By introducing `isWritable` we can get the batch to be drained slightly quicker.

Not sure if the extra state is worth that optimization though.

junrao replied:
Hmm, in RecordAccumulator.append(), we return the following. So if the current batch doesn't have enough space, we will create another one. Then dq.size() will be > 1 and isFull will be true anyway.

`return new RecordAppendResult(future, dq.size() > 1 || batch.isFull(), true);`

apurvam replied:
Yes. That makes sense. I will drop `isWritable`.

Commit:
Addressed most broker side PR comments. Notable changes:

  1. Added a scheduled task to produce snapshots.
  2. Cleaned up the pid snapshot offset management. There was an off by
one error where during truncation we would leave records which should
have been truncated. The reason is that the offsets in the pid mapping
table went from (-1, logEndOffset], inclusive. In the rest of the code,
the offsets range from (0, logEndOffset), exclusive.
  3. Don't add expired Pids to the mapping when loading snapshots.

Commit:
Two changes:

  1. When loading the pid snapshot, we now handle the
CorruptSnapshotException: If we get a corrupt snapshot, we delete it and
move on to the next latest snapshot.
  2. In the Producer, we set the default retries and max inflights
config to be workable for the idempotent producer if idempotence is
enabled.

Commit:
Addressed more PR comments

Commit:
Fix checkStyle error

apurvam commented:
@junrao I addressed most of your comments, except two. Especially for `sendAndAwaitInitPidRequest` comment, I would like to sync face to face tomorrow.

Commit:
Drop isWritable from ProducerBatch since it is redundant.

Commit:
Remove isWritable completely, update a test which is no longer valid,
since we don't close a batch when it is full.

Commit:
Add comment which got deleted

Commit:
Move to CRC32C

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -327,33 +296,100 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+
+    private static int configureInflightRequests(ProducerConfig config, boolean idempotenceEnabled) {
+        int configuredMaxInflightRequests = config.getInt(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION);
+        if (idempotenceEnabled && configuredMaxInflightRequests == ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION_DEFAULT) {
```
A more reliable way to check if a user explicitly sets the config is to check from config.originals(). Ditto for configureRetries().

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -56,6 +56,8 @@ object LogAppendInfo {
  * @param validBytes The number of valid bytes
  * @param offsetsMonotonic Are the offsets in this message set monotonically increasing
+ * @param producerAppendInfos A map from a Pid to a ProducerAppendInfo, which is used to validate each Record in a
+ *                            RecordBatch and keep track of metadata across Records in a RecordBatch.
```
Should we add the comment for isDuplicate too?

apurvam replied:
Added a line for `isDuplicate`

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -163,10 +163,14 @@ object Defaults {
 
   val CompressionType = "producer"
 
+  val MaxIdMapSnapshots = 1
```
Perhaps we should default to 2 to increase the chance that we can rebuild from a snapshot after truncation?

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,390 @@
+  def append(batch: RecordBatch): Unit =
+    append(batch.producerEpoch, batch.baseSequence, batch.lastSequence, batch.maxTimestamp, batch.lastOffset)
+
+  def append(entry: ProducerIdEntry): Unit =
```
It seems this method  is never called?

apurvam replied:
Actually, this is called from `ProducerIdMappingTest.checkAndUpdate`. Not sure why the IDE doesn't find the usage, but the compiler definitely does!

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,390 @@
+  private var firstSeq = initialEntry.firstSeq
+  private var lastSeq = initialEntry.lastSeq
+  private var lastOffset = initialEntry.lastOffset
+  private var lastTimestamp = initialEntry.timestamp
```
lastTimestamp => maxTimestamp? Also, probably add some comments to make it clear that firstSeq, lastSeq lastOffset and lastTimestamp refer to what's in the last appended batch?

apurvam replied:
Changed the name and added a comment.

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,390 @@
+  def isDuplicate(batch: RecordBatch): Boolean = {
+    batch.producerEpoch == epoch &&
+      batch.baseSequence == firstSeq &&
+      batch.lastSequence() == lastSeq
```
Remove () to be consistent with how we call other methods?

junrao reviewed (commented):
@apurvam : Thanks for the patch. LGTM. Just a few more minor comments.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -224,10 +227,16 @@
 
+    /** <code>enable.idempotence</code> */
+    public static final String ENABLE_IDEMPOTENCE_CONFIG = "enable.idempotence";
+    public static final String ENABLE_IDEMPOTENCE_DOC = "When set to 'true', the producer will ensure that exactly one copy of each message is written in the stream. If 'false', producer "
```
We should probably mention in the docs that enabling idempotence will change the default configurations for retries and in-flight requests. I think it might also be worth adding at least a `debug` level log message in the code that we are overriding the defaults.

apurvam replied:
Good point. I think an `info` level log would be even more appropriate.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -222,17 +228,24 @@ public RecordAppendResult append(TopicPartition tp,
 
+    private MemoryRecordsBuilder getRecordsBuilder(ByteBuffer buffer, byte maxUsableMagic) {
+        if (transactionState != null && maxUsableMagic < RecordBatch.MAGIC_VALUE_V2) {
+            throw new UnsupportedVersionException("Attempting to use idempotence with a broker which does not " +
```
Do we have a test case for this?

apurvam replied:
Just added one.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -418,7 +431,27 @@ public boolean hasUnsent() {
+                                            // the pid and sequence here, this attempt will also be accepted, causing
+                                            // a duplicate.
+                                            int sequenceNumber = transactionState.sequenceNumber(batch.topicPartition);
+                                            log.debug("Dest: {} : pid: {}, epoch: {}, Assigning sequence for {}: {}",
```
nit: we might want to spell out "ProducerId" in log messages, so there's no potential for confusion.

apurvam replied:
ok will update as and when I see them. Updated this one.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java:
```diff
@@ -0,0 +1,127 @@
+     * This method is used when the producer needs to reset it's internal state because of an irrecoverable exception
+     * from the broker.
+    */
+    public synchronized void reset() {
```
Given that this class will be used to maintain transactional state as well, perhaps we should use a more explicit name. For example, `resetProducerId`, or maybe `invalideProducerId`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +282,50 @@ public void forceClose() {
+                            "We will back off and try again.");
+                }
+            } catch (Exception e) {
+                log.warn("Received an exception {} while trying to get a pid. Will back off and retry.", e);
```
We shouldn't need a placeholder for the exception unless the intention is to not print the stack trace.

apurvam replied:
In this case, we want to swallow the exception and try again, since we can't do anything without a Pid when idempotence is enabled.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -300,39 +373,85 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
+                                transactionState.sequenceNumber(batch.topicPartition));
+                        reenqueueBatch(batch, now);
+                    } else {
+                        failBatch(batch, response, new OutOfOrderSequenceException("Attempted to retry sending a " +
```
Maybe we can add the PIDs to the exception message?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -300,39 +373,85 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
+                } else {
+                    if (transactionState.pidAndEpoch().producerId == batch.producerId()) {
+                        // If idempotence is enabled only retry the request if the current PID is the same as the pid of the
+                        // batch.
```
nit: move to previous line

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -300,39 +373,85 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
+                if (transactionState == null) {
+                    reenqueueBatch(batch, now);
+                } else {
+                    if (transactionState.pidAndEpoch().producerId == batch.producerId()) {
```
nit: this could be `else if`?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -166,12 +169,15 @@
+    POLICY_VIOLATION(44, new PolicyViolationException("Request parameters do not satisfy the configured policy.")),
+    OUT_OF_ORDER_SEQUENCE_NUMBER(45, new OutOfOrderSequenceException("The broker received an out of order sequence number")),
+    DUPLICATE_SEQUENCE_NUMBER(46, new DuplicateSequenceNumberException("The broker received a duplicate sequence number")),
+    PRODUCER_FENCED(47, new ProducerFencedException("Producer attempted an operation with an old epoch"));
```
We don't have to do it here, but we should make these messages a bit more user-friendly. For example, here we should mention the fact that an old epoch means that this process is probably a zombie and another producer has taken over.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -577,4 +591,11 @@ public RecordsInfo(long maxTimestamp,
     }
+
+    /**
+     * Return the ProducerId (PID) of the LogEntries created by this builder.
```
Replace log entries with record batches

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -327,33 +296,100 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+
+    private static int configureRetries(ProducerConfig config, boolean idempotenceEnabled) {
+        if (idempotenceEnabled && config.getInt(ProducerConfig.RETRIES_CONFIG) == 0) {
+            return 3;
```
I don't have all the context, but isn't `3` pretty low? We don't do exponential back-offs, so the recommendation for no data loss is typically higher.

apurvam replied:
What is the recommendation for no data loss?

ijuma replied:
Great question. :) Infinite is often what is said in talks. But that may not be the right value either. I think it's worth thinking about what the retries help us recover from and how long would we want to keep retrying for. Because with the default retry backoff of 100ms, 3 retries get used pretty fast. Say that we wanted to keep retrying for 10 seconds, that would be roughly 100 retries.

The other side of the coin is: what is the cost of having a high retry number?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/InitPidResponse.java:
```diff
@@ -0,0 +1,75 @@
+
+import java.nio.ByteBuffer;
+
+public class InitPidResponse extends AbstractResponse {
```
Can we list the possible errors in a comment like we do for other responses?

apurvam replied:
Hmm. The only real error codes are when there are transactions, ie. `CoordinatorNotAvailable`, `InvalidTransactionTimeout`, and `NotCoordinatorForTransactionalID`. I wonder if it makes sense to add these in this patch, or wait till we add transactions.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -381,6 +372,150 @@ public void testMetadataTopicExpiry() throws Exception {
+        }, new InitPidResponse(Errors.NONE, producerId, (short) 0));
+        sender.run(time.milliseconds());
+        assertTrue(transactionState.hasPid());
+        assertEquals(transactionState.pidAndEpoch().producerId, producerId);
```
We can assert the epoch also? Also, this is backwards: the expected value should be listed first.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -381,6 +372,150 @@ public void testMetadataTopicExpiry() throws Exception {
+    }
+
+    @Test
+    public void testSequenceNumberIncrement() throws InterruptedException {
```
I was looking for a test case which verified that the producer PID, epoch, and sequence number are set correctly in the records included with the produce request (e.g. using a `RequestMatcher`). Do we have one?

apurvam replied:
I just added this test.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -381,6 +372,150 @@ public void testMetadataTopicExpiry() throws Exception {
+    }
+
+    @Test
+    public void testAbortWhenPidChanges() throws InterruptedException {
```
What are we aborting? Can you clarify in the name?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionStateTest.java:
```diff
@@ -0,0 +1,50 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class TransactionStateTest {
```
Missing a test case for `reset`?

apurvam replied:
Added a test case.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -787,6 +790,14 @@ private DeleteTopicsResponse createDeleteTopicsResponse() {
         return new DeleteTopicsResponse(errors);
     }
 
+    private InitPidRequest createInitPIDRequest() {
```
nit: the convention we're using elsewhere is `Pid`. Same below.

hachikuji commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -196,7 +195,7 @@ class RequestSendThread(val controllerId: Int,
               val clientRequest = networkClient.newClientRequest(brokerNode.idString, requestBuilder,
                 time.milliseconds(), true)
-              clientResponse = networkClient.blockingSendAndReceive(clientRequest)(time)
+              clientResponse = NetworkClientUtils.sendAndReceive(networkClient, clientRequest, time)
```
nit: we could probably `import NetworkClientUtils._`

apurvam replied:
I actually prefer to avoid wildcard imports.

ijuma replied:
Hmm, but this same file has been changed to use wildcard imports at the top (e.g. `import org.apache.kafka.common.network._`)?

hachikuji commented on core/src/main/scala/kafka/coordinator/PidMetadata.scala:
```diff
@@ -0,0 +1,31 @@
+private[coordinator] class PidMetadata(val pid: Long) {
+
+  /* current epoch number of the PID */
+  var epoch: Short = 0
```
Might be a good idea to keep this private. We could add an `increment` method instead of writing to the field directly from external classes.

hachikuji commented on core/src/main/scala/kafka/coordinator/ProducerIdManager.scala:
```diff
@@ -0,0 +1,153 @@
+
+class ProducerIdManager(val brokerId: Int, val zkUtils: ZkUtils) extends Logging {
+
+  this.logIdent = "[PID Manager " + brokerId + "]: "
```
nit: can we make this `ProducerId Manager`?

hachikuji commented on core/src/main/scala/kafka/coordinator/ProducerIdManager.scala:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  private def getNewPidBlock(): Unit = {
+    var zkWriteComplete: Boolean = false
```
nit: no need for the type on the left-hand side.

hachikuji commented on core/src/main/scala/kafka/coordinator/TransactionCoordinator.scala:
```diff
@@ -0,0 +1,92 @@
+    if (transactionalId == null || transactionalId.isEmpty) {
+      // if the transactional id is not specified, then always blindly accept the request
+      // and return a new pid from the pid manager
+      val pid: Long = pidManager.nextPid()
```
nit: no need for type on lhs.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -332,6 +357,46 @@ class Log(@volatile var dir: File,
+            // or magic causes us to load the full entry into memory. It would be better if we
+            // only loaded the header
+            val pidEntry = ProducerIdEntry(batch.producerEpoch, batch.lastSequence, batch.lastOffset,
+              (batch.lastOffset - batch.baseOffset + 1).toInt, batch.maxTimestamp)
```
nit: we can assign `batch.lastoffset - batch.baseOffset +1` to a local variable so it's easier to understand.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -434,13 +499,23 @@ class Log(@volatile var dir: File,
 
 
-        // now append to the log
+        // update the first offset if on the first message. For magic versions older than 2, we use the last offset
```
Why was the comment moved here?

apurvam replied:
Not sure how that moved around like that.  I think it must have been a fat fingered cut/paste from last night. Moved it back now.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -488,7 +563,7 @@ class Log(@volatile var dir: File,
    * </ol>
    */
-  private def analyzeAndValidateRecords(records: MemoryRecords): LogAppendInfo = {
+  private def analyzeAndValidateRecords(records: MemoryRecords, isFromClient: Boolean): LogAppendInfo = {
```
Probably we should add something to the comment above about doing PID validation.

apurvam replied:
Added the comment.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -527,19 +603,41 @@ class Log(@volatile var dir: File,
+          case Some(appendInfo) => appendInfo.append(batch)
+          case None =>
+            val lastEntry = pidMap.lastEntry(pid).getOrElse(ProducerIdEntry.Empty)
+            if (isFromClient && lastEntry.isDuplicate(batch)) {
```
Might be worth mentioning that the loop will only iterate once for a duplicate client request to be clear that the values below will not be overwritten. Would be nice if we could just break, but alas.

apurvam replied:
Added the comment.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -527,19 +603,41 @@ class Log(@volatile var dir: File,
+              firstOffset = lastEntry.firstOffset
+              lastOffset = lastEntry.lastOffset
+              maxTimestamp = lastEntry.timestamp
+              warn(s"Detected a duplicate at (firstOffset, lastOffset): (${firstOffset}, ${lastOffset}). Ignoring the incoming record.")
```
`warn` seems a bit high. Could this be `debug`?

apurvam replied:
I'll bump it down to info. debug seems too low, since this should happen fairly rarely.

hachikuji commented on core/src/main/scala/kafka/log/LogConfig.scala:
```diff
@@ -127,6 +131,9 @@ object LogConfig {
   val FollowerReplicationThrottledReplicasProp = "follower.replication.throttled.replicas"
 
+  /** ********* Id mapping snapshots ****************/
+  val MaxIdMapSnapshotsProp = "max.id.map.snapshots"
```
Is this config part of the KIP? Do we really need it? I know we had discussed at one point just using a reasonable default.

apurvam replied:
Hmm. It is not part of the KIP. We agreed that 2 would be a good static value, I think. Will update.

hachikuji commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,390 @@
+    }
+  }
+
+  def isEntryValid(currentTimeMs: Long, producerIdEntry: ProducerIdEntry) : Boolean = {
```
Not clear what a "valid" entry is. Maybe the check should be inverted and phrased as `hasEntryExpired`?

apurvam replied:
The reason to choose a positive name is that we use it to retain unexpired entries. If we choose something like `hasEntryExpired` we would have to negate the usage everywhere.

Seems like both have their tradeoffs, so I would prefer to leave it as is.

hachikuji commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,390 @@
+
+    var loaded = false
+    while (!loaded) {
+      lastSnapshotFile(logEndOffset) match {
```
Seems this loop would be a little clearer if we just built a list of the snapshot files sorted by offset and iterated over it?

apurvam replied:
Not sure if it would be more clear, but probably more efficient. I made the change this way so that we retain the existing logic for picking the latest snapshot less than the given offset.

hachikuji commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -0,0 +1,390 @@
+    *
+    *  The snapshot files are located in the logDirectory, inside a 'pid-mapping' sub directory.
+    */
+  def maybeTakeSnapshot() {
```
One thing I was thinking about (for another patch) is whether we should have like a minimum number of messages before it's worth doing another snapshot. If only 5 messages have been written since the last snapshot, for example, maybe we can just skip the new snapshot.

hachikuji commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -119,7 +119,7 @@ abstract class AbstractFetcherThread(name: String,
       case t: Throwable =>
         if (isRunning.get) {
-          warn(s"Error in fetch $fetchRequest", t)
+          warn(s" fetch $fetchRequest", t)
```
Was this intentional?

apurvam replied:
Nope. seems to have been there since fpj's time. I fixed it.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -179,9 +182,9 @@ class KafkaApis(val requestChannel: RequestChannel,
-        result.foreach { case (topicPartition, error) =>
-          if (error == Errors.NONE && stopReplicaRequest.deletePartitions() && topicPartition.topic == Topic.GroupMetadataTopicName) {
-            coordinator.handleGroupEmigration(topicPartition.partition)
+        result.foreach { case (topicPartition, errorCode) =>
```
Not sure why the variable name was changed: `error` is more accurate.

hachikuji commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -120,6 +120,219 @@ class LogTest extends JUnitSuite {
+
+    log.append(MemoryRecords.readableRecords(buffer), assignOffsets = false)
+    // Should throw a duplicate seqeuence exception here.
+    assertFalse("should have thrown a DuplicateSequenceNumberException.", true)
```
nit: you can use `fail` instead

hachikuji commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -42,6 +42,7 @@ def __init__(self, test_context):
         # producer begins producing messages, in which case we will miss the
         # initial set of messages and get spurious test failures.
         self.consumer_init_timeout_sec = 0
+        self.enable_idempotence=False
```
nit: seems convention is to have spaces before and after `=`?

hachikuji reviewed (commented):
Thanks for the PR. Left a bunch of minor comments.

ijuma commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -449,9 +448,16 @@ private[log] class Cleaner(val id: Int,
+        // the PID is not removed from the log before it has been expired
+        val pid = recordBatch.producerId
+        val isLastEntryForPid = RecordBatch.NO_PRODUCER_ID < pid && activePids.get(pid).exists(_.lastOffset == record.offset)
+        record.isControlRecord || isLastEntryForPid || shouldRetainMessage(source, map, retainDeletes, record, stats)
```
I don't think we should split the logic between this and `shouldRetainMessage`. For example, the latter already checks `record.isControlRecord`.

apurvam replied:
I consolidated all the logic into `shouldRetainMessage`.

Commit:
Addressed remaining PR comments

apurvam commented:
Latest run of the system tests on this branch succeeded: https://jenkins.confluent.io/job/idempotent-producer-system-tests/5/console

Kicked off a muckrake run against this branch too : http://jenkins.confluent.io/job/system-test-confluent-platform-branch-builder/203

Commit:
Consolidate all the retention logic inside LogCleaner.shouldRetainMessage

apurvam commented:
Thanks for the review @junrao, @hachikuji, and @ijuma .. I think I have addressed all the comments. The system tests are running as well (links above).

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +282,50 @@ public void forceClose() {
+        return NetworkClientUtils.sendAndReceive(client, request, time);
+    }
+
+    private Node getReadyNode(long remainingTimeMs) throws IOException {
```
Wouldn't `awaitLeastLoadedNodeReady` be a clearer name?

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java:
```diff
@@ -194,6 +194,36 @@ public Integer validate(Object item) {
+        }
+
+        @Override
+        public Object read(ByteBuffer buffer) {
```
Nit: it's a bit nicer if we return `Long` here.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -1176,6 +1176,29 @@
+            new Field("error_code",
+                    INT16,
+                    "An integer error code."),
+            new Field("pid",
```
For my benefit, when we do we use `producer_id` and when do we use `pid`?

apurvam replied:
I think it is a bit arbitrary. It started with pid everywhere, but then started changing gradually to producerId or producer_id.

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -338,13 +338,27 @@ public static MemoryRecords withRecords(long initialOffset, CompressionType comp
         ByteBuffer buffer = ByteBuffer.allocate(sizeEstimate);
-        MemoryRecordsBuilder builder = builder(buffer, magic, compressionType, timestampType, initialOffset);
+        MemoryRecordsBuilder builder = builder(buffer, magic, compressionType, timestampType, initialOffset,
+                System.currentTimeMillis(), pid, epoch, baseSequence);
```
We can avoid the `currentTimeMillis` if `timestampType` is `CREATE_TIME`. In that case, we can simply use `NO_TIMESTAMP`. The same applies for a couple of other methods.

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -193,6 +193,20 @@ else if (maxTimestamp == RecordBatch.NO_TIMESTAMP)
+            // be re queued. In this case, we should not attempt to set the state again, since changing the pid and sequence
+            // once a batch has been sent to the broker risks introducing duplicates.
+            throw new IllegalStateException("Attempting to close the same batch with a different base sequence. "
+                    + "Being in this situation indicates data corruption is afoot.");
```
It would probably be useful to include some data in this message. Maybe the existing pid/epoch/baseSequence and the new proposed values?

apurvam replied:
Actually, the whole error message is out dated. We should not be calling `setProducerState` on a closed batch any more. Doing so is a bug on the client. Updated the message to indicate that.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java:
```diff
@@ -31,6 +31,16 @@
+     * @param buffer The buffer to read from
+     * @return The integer read, as a long to avoid signedness
+     */
+    public static long readUnsignedInt(ByteBuffer buffer) {
```
Please add a simple unit test in ByteUtilsTest for this.

apurvam replied:
Done.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -519,6 +523,17 @@ public void testMutedPartitions() throws InterruptedException {
+        // Simulate talking to an older broker, ie. one which supports a lower magic.
+        ApiVersions apiVersions = new ApiVersions();
+        int batchSize = 1025;
+        apiVersions.update("foobar", NodeApiVersions.create(Arrays.asList(new ApiVersionsResponse.ApiVersion(ApiKeys.PRODUCE.id, (short) 0, (short) 2))));
```
Nit: long line.

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -624,6 +624,20 @@ class ZkUtils(val zkClient: ZkClient,
+      if (data == null.asInstanceOf[String])
+        (None, stat.getVersion)
+      else
+      (Some(data), stat.getVersion)
```
The 5 lines above can be written as:

```scala
val data: String = zkClient.readData(path, stat)
(Option(data), stat.getVersion)
```

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -719,6 +733,14 @@ class ZkUtils(val zkClient: ZkClient,
+    if (topicData(topic).nonEmpty)
+      Some(topicData(topic).size)
+    else
+      None
```
Simpler perhaps:

```scala
val topicData = getPartitionAssignmentForTopics(Seq(topic))
Some(topicData(topic).size).filter(_ > 0)
```

ijuma commented on core/src/test/scala/unit/kafka/utils/ZkUtilsTest.scala:
```diff
@@ -40,6 +40,24 @@ class ZkUtilsTest extends ZooKeeperTestHarness {
   }
 
+  /*
+   * Added this test since PIDManager's behavior is depending on the expected return values
```
Nit: it seems odd to have "Added this test" as a test comment. Something like "Verify behaviour of ZkUtils.createSequentialPersistentPath since PIDManager relies on it" seems like the expected style.

ijuma commented on core/src/test/scala/unit/kafka/server/ReplicationQuotasTest.scala:
```diff
@@ -201,6 +201,7 @@ class ReplicationQuotasTest extends ZooKeeperTestHarness {
     val start = System.currentTimeMillis()
 
     //Start the new broker (and hence start replicating)
+    error("Starting new broker")
```
Why do we have this at error level?

apurvam replied:
Was used during debugging. reverted to debug level.

ijuma reviewed (commented):
A few minor comments.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -210,9 +227,21 @@ void run(long now) {
-        for (ProducerBatch expiredBatch : expiredBatches)
+
+        boolean needsTransactionStateReset = false;
+        // Reset the PID if an expired batch has previously been sent to the broker. Also update the metrics
```
Nit: I personally find comments that just repeat what the code is doing not so useful. However, if we explained why we need to reset the pid in this case, that would be pretty useful.

apurvam replied:
I added this documentation to the `TransactionState.resetProducerId` method.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -222,17 +228,24 @@ public RecordAppendResult append(TopicPartition tp,
-     * resources (like compression streams buffers).
+     *  Try to append to a ProducerBatch. If it is full, we return null and a new batch is created. If the existing batch is
+     *  full, it will be closed right before send, or if it is expired, or when the producer is closed, whichever
+     *  comes first.
```
Have we done any performance tests to see the impact of this change? The change I made to close the memory records here made a huge difference to the amount of memory used by the producer due to temporary compression buffers.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -222,17 +228,24 @@ public RecordAppendResult append(TopicPartition tp,
         }
     }
 
+    private MemoryRecordsBuilder getRecordsBuilder(ByteBuffer buffer, byte maxUsableMagic) {
```
Nit: this could just be `recordsBuilder`. We generally avoid the `get` prefix in Kafka.

Commit:
Fixed a bug where when records are compressed and the record has to be
rewritten on the server, we would lose the pid, sequence, and epoch of
the record batch.

Updated the integration tests to include compressed records. Without
this fix, the integration test would fail.

Could not yet reproduce using unit tests. Will continue working on that
and add it eventually.

Commit:
Address more PR comments

apurvam commented:
The jenkins builds are puzzling. Mostly different streams tests fail in each run. There hasn't been a failure in core or clients for a while.

The perf numbers are being collected here: https://docs.google.com/spreadsheets/d/1dHY6M7qCiX-NFvsgvaE0YoVdNq26uA8608XIh_DUpI4/edit#gid=1276994626

So far, so good. Actually enabling the idempotent producer costs 20% throughput. But using the new code with idempotent producer turned off seems to have no effect. 

A muckrake test is finally running properly and seems to be passing: http://jenkins.confluent.io/job/system-test-confluent-platform-branch-builder/206/console

I also addressed most of the major comments from Ismael, ie. those which had to do with adding test cases or improving error messages or improving documentation. The remaining ones are more to with minor code style.

apurvam commented:
retest this please

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -327,33 +296,112 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+        }
+        if (idempotenceEnabled && !userConfiguredRetries) {
+            log.info("Overriding the default retries config to " + 3 + " since the idempotent producer is enabled.");
+            return 3;
```
Similar to this, it seems the default acks=1 doesn't make sense when idempotence is enabled. This is because with acks=1, acked messages could be lost during leader change. Then, the producer will be out of sequence. Perhaps if idempotence is enabled, we should enforce acks=all.

apurvam replied:
Sounds good. Of course, we also depend on some topic level settings like: replication.factor >= 3, min.isr >= 2, unclean.leader.election=false.

But I agree, we should do what we can on the client.

apurvam replied:
I added the override for acks. Will make a note to update the KIP as well.

apurvam commented:
Muckrake test passed: http://jenkins.confluent.io/job/system-test-confluent-platform-branch-builder/206/

System test against this branch also passed: https://jenkins.confluent.io/job/idempotent-producer-system-tests/7/

Investigating the jenkins branch builder failures.

Commit:
Set the default acks to all when idempotent producer is enabled. If the configured value is not all and idempotent producer is set, we throw a config exception

apurvam commented:
Latest failure (JDK8 and Scala 2.11): 

```
org.apache.kafka.streams.StreamsConfigTest > shouldSupportPrefixedRestoreConsumerConfigs PASSED
pure virtual method called
terminate called without an active exception
:streams:unitTest FAILED
```
Somehow my local runs are super stable. I don't know what would cause this kind of thing.

junrao commented:
@apurvam : Thanks for the patch. LGTM

junrao commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -493,10 +493,17 @@ private[log] class Cleaner(val id: Int,
     if (record.isControlRecord)
       return true
 
+    // retain the entry if it is the last one produced by an active idempotent producer to ensure that
```
entry => record

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -497,8 +574,12 @@ class Log(@volatile var dir: File,
 
     for (batch <- records.batches.asScala) {
+      if (isFromClient && batch.magic >= RecordBatch.MAGIC_VALUE_V2 && shallowMessageCount > 0)
+        throw new InvalidRecordException("Client produce requests should not have more than one batch")
```
I thought we agreed that this check is unnecessary given the check in ProduceRequest?

apurvam replied:
We did, initially, but then I followed up on that thread (which is now impossible to find on github). Here is what I wrote: 

> So I thought about this a bit more. I think it still make sense to validate this on the server side. I assume that the librdkafka clients may not have the request side validation, so it would be good to have server side validation before we write to the log.

junrao replied:
Hmm, ProducerRequest.validateRecords() is called on the broker side when converting bytes from socket to a request object. So, even if librd client has an issue, the broker should still be able to capture this when constructing ProducerRequest.

junrao reviewed (commented):
A couple of minor comments. Probably you can address them in a followup patch.
